cpp/src/arrow/adapters/orc/adapter_util.cc

```diff
@@ -315,13 +344,662 @@ Status AppendBatch(const liborc::Type* type, liborc::ColumnVectorBatch* batch,
+  RETURN_NOT_OK(visitor.Visit(*(array_.data()), &appender));
+  return arrow::Status::OK();
+}
+
+arrow::Status WriteFixedSizeBinaryBatch(const arrow::Array& array, int64_t orc_offset,
+                                        liborc::ColumnVectorBatch* column_vector_batch) {
+  const arrow::FixedSizeBinaryArray& array_(
+      checked_cast<const arrow::FixedSizeBinaryArray&>(array));
+  auto batch = checked_cast<liborc::StringVectorBatch*>(column_vector_batch);
+  if (array.null_count()) {
+    batch->hasNulls = true;
+  }
+  FixedSizeBinaryAppender appender{array_, batch, orc_offset, 0, array_.byte_width()};
+  arrow::ArrayDataVisitor<arrow::FixedSizeBinaryType> visitor;
+  RETURN_NOT_OK(visitor.Visit(*(array_.data()), &appender));
+  return arrow::Status::OK();
+}
+
+arrow::Status WriteStructBatch(const arrow::Array& array, int64_t orc_offset,
+                               liborc::ColumnVectorBatch* column_vector_batch,
+                               bool normalized) {
+  std::shared_ptr<arrow::Array> array_ = arrow::MakeArray(array.data());
+  std::shared_ptr<arrow::StructArray> struct_array(
+      std::static_pointer_cast<arrow::StructArray>(array_));
+  auto batch = checked_cast<liborc::StructVectorBatch*>(column_vector_batch);
+  std::size_t size = array.type()->fields().size();
+  int64_t arrow_length = array.length();
+  int64_t running_arrow_offset = 0, running_orc_offset = orc_offset;
+  // First fill fields of ColumnVectorBatch
+  if (array.null_count()) {
+    batch->hasNulls = true;
+  }
+  for (; running_arrow_offset < arrow_length;
+       running_orc_offset++, running_arrow_offset++) {
+    if (array.IsNull(running_arrow_offset)) {
+      batch->notNull[running_orc_offset] = false;
+    } else {
+      batch->notNull[running_orc_offset] = true;
+    }
+  }
+  // Fill the fields
+  for (std::size_t i = 0; i < size; i++) {
+    batch->fields[i]->resize(orc_offset + arrow_length);
+    RETURN_NOT_OK(
+        WriteBatch(*(struct_array->field(i)), orc_offset, batch->fields[i], true));
+  }
+  return arrow::Status::OK();
+}
+
+template <class ArrayType>
+arrow::Status WriteListBatch(const arrow::Array& array, int64_t orc_offset,
+                             liborc::ColumnVectorBatch* column_vector_batch) {
+  const ArrayType& list_array(checked_cast<const ArrayType&>(array));
+  auto batch = checked_cast<liborc::ListVectorBatch*>(column_vector_batch);
+  liborc::ColumnVectorBatch* element_batch = (batch->elements).get();
+  int64_t arrow_length = array.length();
+  int64_t running_arrow_offset = 0, running_orc_offset = orc_offset;
+  if (orc_offset == 0) {
+    batch->offsets[0] = 0;
+  }
+  if (array.null_count()) {
+    batch->hasNulls = true;
+  }
+  for (; running_arrow_offset < arrow_length;
+       running_orc_offset++, running_arrow_offset++) {
+    if (array.IsNull(running_arrow_offset)) {
+      batch->notNull[running_orc_offset] = false;
+      batch->offsets[running_orc_offset + 1] = batch->offsets[running_orc_offset];
```

Comment:
You must also take into account `list_array.value_offset()` just like in the non-null case. Some list arrays might have null entries with a non-zero length.

Reply:
Please correct me if I'm wrong: I know that they do and unlike Arrow ORC is not supposed to accommodate this behavior since we prioritize saving space as opposed to guaranteeing O(1) access in ORC. This is why I deliberately use `batch->offsets[running_orc_offset + 1] = batch->offsets[running_orc_offset];` to skip these elements that shouldn't make their way into the ORC elements array.